bananas/drf/fencing.py

```diff
@@ -0,0 +1,181 @@
+from drf_yasg.inspectors import SwaggerAutoSchema
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework.mixins import UpdateModelMixin
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.serializers import BaseSerializer, ModelSerializer
+from rest_framework.viewsets import GenericViewSet
+from typing_extensions import Final, final
+
+from bananas.models import TimeStampedModel
+
+from . import errors
+from .utils import HeaderError, parse_header_datetime, parse_header_etags
+
+__all__ = (
+    "Fence",
+    "FencedUpdateModelMixin",
+    "header_date_parser",
+    "parse_date_modified",
+    "allow_if_unmodified_since",
+    "header_etag_parser",
+    "allow_if_match",
+)
+
+TokenType = TypeVar("TokenType")
+InstanceType = TypeVar("InstanceType")
+
+
+@final
+class Fence(abc.ABC, Generic[InstanceType, TokenType]):
+    def __init__(
+        self,
+        get_token: Callable[[Request], TokenType],
+        compare: Callable[[TokenType, TokenType], bool],
+        get_version: Callable[[InstanceType], Optional[TokenType]],
+        openapi_parameter: openapi.Parameter,
+    ) -> None:
+        self._get_token: Final = get_token
+        self._compare: Final = compare
+        self._get_version: Final = get_version
+        self.openapi_parameter: Final = openapi_parameter
+
+    def check(self, request: Request, instance: InstanceType) -> bool:
+        version = self._get_version(instance)
+        if version is None:
+            # We might want to expose control of this behavior. For if-unmodified-since
+            # it makes sense to return true here, but it might not for
+            # if-modified-since other conditionals.
+            return True
+        return self._compare(version, self._get_token(request))
+
+
+class FenceAwareSwaggerAutoSchema(SwaggerAutoSchema):
+    update_methods = ("PUT", "PATCH")
+
+    def add_manual_parameters(
+        self, parameters: List[openapi.Parameter]
+    ) -> List[openapi.Parameter]:
+        parameters = super().add_manual_parameters(parameters)
+        if (
+            isinstance(self.view, FencedUpdateModelMixin)
+            and self.method in self.update_methods
+        ):
+            return parameters + [self.view.fence.openapi_parameter]
+        return parameters
+
+
+class FencedUpdateModelMixin(UpdateModelMixin, abc.ABC):
+    @property
+    @abc.abstractmethod
+    def fence(self) -> Fence:
+        ...
+
+    def perform_update(self, serializer: BaseSerializer) -> None:
+        # mypy's advanced self types don't work with super calls so we use an assertion
+        # here instead.
+        assert isinstance(self, GenericViewSet)
+        assert isinstance(serializer, ModelSerializer)
+        if not self.fence.check(self.request, serializer.instance):
+            raise errors.PreconditionFailed(
+                "The resource does not fulfill the given preconditions"
+            )
```

Comment:
The raised error is specific to the fence, for instance `If-Modified-Since` should return a 304 when rejected, so I went with exposing a `rejection` parameter. What do you think? :)

Reply:
That makes sense!